src/main/scala/ru/otus/sc/game/dao/impl/GameProcessDaoImpl.scala

```diff
@@ -0,0 +1,347 @@
+package ru.otus.sc.game.dao.impl
+
+import ru.otus.sc.game.dao.{GameProcessDao, GameStoreDao}
+import ru.otus.sc.game.model.GameProcessState._
+import ru.otus.sc.game.model._
+
+import scala.collection.mutable
+import scala.util.Random
+
+class GameProcessDaoImpl(gsd: GameStoreDao) extends GameProcessDao {
+  private val gameState: mutable.Map[String, PlayerCurrentGameState] = mutable.Map.empty
+  private val size: Int                                              = GameStats.worldSize
+  private val rand                                                   = new Random(size)
+
+  @scala.annotation.tailrec
+  private def freeCell(map: Map[Position, GameEntity]): Position = {
+    val pos = Position(rand.nextInt(size), rand.nextInt(size))
+    if (!map.contains(pos))
+      pos
+    else
+      freeCell(map)
+  }
+
+  /**
+    * Получение игрового состояния игрока
+    *
+    * @param player игрок
+    * @return игровое состояние игрока
+    */
+  override def gamerState(player: Player): GameProcessState = {
+    this.gameState.get(player.nick) match {
+      case Some(playerGame) =>
+        playerGame.state
+      case None =>
+        GameProcessState.NONE
+    }
+  }
+
+  /**
+    * Игрок начинает новую игру.
+    * Должна быть создана новая карта и размещены юниты
+    *
+    * @param player игрок
+    */
+  override def newGame(player: Player): Boolean = {
+    val localMap = GameMap(3)
+    val mapWithEnemies = GameMap(
+      List
+        .range(0, 5)
+        .foldLeft(localMap.entities)((acc, _) => {
+          acc + (freeCell(acc) -> Player.enemy())
+        })
+    )
+    val pos = freeCell(mapWithEnemies.entities)
+    this.gameState.get(player.nick) match {
+      case Some(gameState) =>
+        this.gameState(player.nick) = gameState.copy(
+          player = player.copy(health = PlayerStats.maxHealth),
+          map = mapWithEnemies.entities,
+          state = GameProcessState.MAIN_MENU,
+          pos = pos
+        )
+      case None =>
+        this.gameState.put(
+          player.nick,
+          PlayerCurrentGameState(player, GameProcessState.MAIN_MENU, mapWithEnemies.entities, pos)
+        )
+    }
+
+    val state = this.gameState(player.nick)
+    this.gsd.save(state.player, state.map, state.pos)
+
+    true
+  }
+
+  /**
+    * Загрузка игры
+    *
+    * @param player игрок
+    * @return смог или не смог
+    */
+  override def loadGame(player: Player): Boolean = {
+    this.gameState.get(player.nick) match {
+      case Some(gameState) =>
+        val savedMap    = this.gsd.loadMap(player)
+        val savedPlayer = this.gsd.loadPlayer(player)
+
+        if (savedMap.isDefined && savedPlayer.isDefined) {
+          this.gameState(player.nick) = gameState.copy(
+            player = savedPlayer.get._1,
+            pos = savedPlayer.get._2,
+            map = savedMap.get
+          )
+          true
+        } else {
+          false
+        }
```

Comment:
```suggestion
        val result = for {
          sMap    <- savedMap
          sPlayer <- savedPlayer
          _        = this.gameState(player.nick) = gameState.copy(sPlayer._1, sPlayer._2, sMap)
        } yield true
        result.getOrElse(false)
```

Опять же чтобы избавиться от `.isDefined`/`get`. На этот раз с помощью `flatMap` (for-comprehension это просто синтаксический сахар для неё)

Можно даже расширить этот for-comprehension до `this.gameState.get(player.nick)`